C++ API parity: Hardtanh

pbelevich opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27309 C++ API parity: MultiheadAttention
* #27382 C++ API parity: Linear
* #27060 C++ API parity: LogSigmoid
* #27059 C++ API parity: LeakyReLU
* **#27038 C++ API parity: Hardtanh**

Differential Revision: [D17682405](https://our.internmc.facebook.com/intern/diff/D17682405)

pbelevich committed:
C++ API parity: Hardtanh

[ghstack-poisoned]

yf225 commented on torch/csrc/api/include/torch/nn/functional/activation.h:
```diff
@@ -20,6 +20,15 @@ inline Tensor hardshrink(const Tensor& input,
 
+inline Tensor hardtanh(const Tensor& input, const HardtanhOptions& options) {
+  if (options.inplace()) {
+    return torch::hardtanh_(const_cast<Tensor&>(input),
```
same here: we might want to just change `input`'s type to `Tensor&`, and remove `const_cast`.

yf225 commented on torch/csrc/api/src/nn/modules/activation.cpp:
```diff
@@ -36,5 +36,23 @@ void HardshrinkImpl::pretty_print(std::ostream& stream) const {
+  stream << std::boolalpha
+         << "torch::nn::Hardtanh(min_val=" << options.min_val()
+         << ", max_val=" << options.max_val()
+         << ", inplace=" << options.inplace() << ")";
```
same here: whether to print `inplace` is conditional on whether it is `true`, in order to match the Python version.

yf225 commented on torch/csrc/api/include/torch/nn/options/activation.h:
```diff
@@ -26,5 +26,21 @@ struct HardshrinkOptions {
+  // maximum value of the linear region range. Default: 1
+  TORCH_ARG(double, max_val) = 1.0;
+
+  // can optionally do the operation in-place. Default: False
```
nit: need `///` instead of `//` (similarly for other PRs)

pbelevich committed:
Update on "C++ API parity: Hardtanh"




[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Hardtanh"




[ghstack-poisoned]

yf225 reviewed (approved):
Thanks @pbelevich !

yf225 commented on torch/csrc/api/src/nn/modules/activation.cpp:
```diff
@@ -38,5 +38,26 @@ void HardshrinkImpl::pretty_print(std::ostream& stream) const {
+// ============================================================================
+
+HardtanhImpl::HardtanhImpl(const HardtanhOptions& options_)
+    : options(options_) {}
```
oops sorry I think we are missing one assert here:
```python
assert self.max_val > self.min_val  # Python version
```

pbelevich committed:
Update on "C++ API parity: Hardtanh"


Differential Revision: [D17682405](https://our.internmc.facebook.com/intern/diff/D17682405)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Hardtanh"


Differential Revision: [D17682405](https://our.internmc.facebook.com/intern/diff/D17682405)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Hardtanh"


Differential Revision: [D17682405](https://our.internmc.facebook.com/intern/diff/D17682405)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Hardtanh"


Differential Revision: [D17682405](https://our.internmc.facebook.com/intern/diff/D17682405)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: Hardtanh"


Differential Revision: [D17682405](https://our.internmc.facebook.com/intern/diff/D17682405)

[ghstack-poisoned]

yf225 commented on torch/csrc/api/src/nn/modules/activation.cpp:
```diff
@@ -38,5 +38,29 @@ void HardshrinkImpl::pretty_print(std::ostream& stream) const {
+HardtanhImpl::HardtanhImpl(const HardtanhOptions& options_)
+    : options(options_) {
+  TORCH_CHECK(options.max_val() > options.min_val(),
+              "max_val must be greater than min_val");
```
Copying from our discussion: I think ideally we should put the check in module reset(), with the expectation that in the future we will support the use case of changing a module’s options and then call reset() to reinitialize the module (without calling its constructor)

yf225 replied:
We need to call reset() from the constructor, if we make that change :)

pbelevich replied:
fixed

pbelevich committed:
Update on "C++ API parity: Hardtanh"


Differential Revision: [D17682405](https://our.internmc.facebook.com/intern/diff/D17682405)

[ghstack-poisoned]

yf225 reviewed (approved):
Thanks a lot @pbelevich !

pbelevich committed:
Update on "C++ API parity: Hardtanh"


Differential Revision: [D17682405](https://our.internmc.facebook.com/intern/diff/D17682405)

[ghstack-poisoned]
